README.md
# Nordnet nExt API Javascript client

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Dependency Status][depstat-image]][depstat-url]

Isomorphic JS client for [Nordnet nExt API][api]. Client can be used both on the client and server side. Client should be used for making HTTP requests towards nExt API. See [nExt API documentation][api-docs] for a list of possible requests.


## Installation

### NPM

```sh
npm install --save nordnet-next-api
```

## Usage

Library can be used on the client and server side.

```js
import api from 'nordnet-next-api';

api
  .get('https://api.test.nordnet.se/next/2')
  .then(({ status, data }) => console.log(status, data));
```

```js
var api = require('nordnet-next-api');
var express = require('express');
var app = express();

app.get('/', function (req, res) {
  api.get('https://api.test.nordnet.se/next/2')
    .then(function(response) {
      res.send(response);
  })
});
```

Authentication is required to be able to use nExt API. `Authorization` header can be used to pass session token when communicating to the API.
See [nExt API documentation][api] for more details on how to get test account and authenticate against nExt API.

### This lib does not use babel-polyfill, so you might have to add it yourself, e.g. if you are building for IE11.

In you application:
```js
import 'babel-polyfill'; // at your application entry point
```

Or in webpack.config.js
```js
{
  entry: ['babel-polyfill', 'your_app_entry.js']
}
```

## API

* `api.get(url, params = {}, headers = {})`
* `api.post(url, params = {}, headers = {})`
* `api.postJson(url, params = {}, headers = {})` — `api.post` with `{ 'Content-type': 'application/json;' }` in headers
* `api.put(url, params = {}, headers = {})`
* `api.putJson(url, params = {}, headers = {})` — `api.put` with `{ 'Content-type': 'application/json;' }` in headers
* `api.del(url, params = {}, headers = {})`

Each method returns a Promise, which resolves or rejects with `Object { response, data, status }` where

* `response`, Type `Object`, [Fetch API Response](https://developer.mozilla.org/en-US/docs/Web/API/Response)
* `data`, Type `Object || String || (undefined if HTTP status === 204)`
* `status`, Type `Number`, [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/API/Response/status)

Promise is rejected when HTTP status code is greater or equal 400.

#### url

*Required*  
Type: `String`
Example:

* `/api/2/login`
* `/api/2/accounts/{accno}`
* `/api/2/instruments/{instrument_id}?positions={positions}`

**`Note:` interpolated url params are taken from `params` argument. If `url` contains a key,
which doesn't exist in `params`, promise will be rejected with `Error`.**

#### params

*Required*  
Type: `Object`  
Default: `{}`

Object `params` is used to
* interpolate `url` params.
* if `headers` contains `"Content-type": "application/json"` for constructing request `payload`.
* otherwise for constructing request body.

#### headers

*Required*  
Type: `Object`  
Default: `{}`

See [Fetch API Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers)



### Basic usage

```js
import api from 'nordnet-next-api';

api
  .get('https://api.test.nordnet.se/next/2/accounts/{accno}', { accno: 123456789 })
  .then(({ status, data, response }) => console.log(status, data, response));
```

Returned response contains
* `status` (HTTP response status)
* `data` (either JSON or plain string depending on `Content-type` header)
* `response` ([Response][response] interface of [Fetch API][fetch-api])

### Setting root URL before querying

```js
import api from 'nordnet-next-api';

api.setConfig({ root: 'https://api.test.nordnet.se/next/2' });

api.get('/accounts/{accno}', { accno: 123456789 })
  .then(response => console.log(response));
```

The following config keys are supported:
* `root` sets base root URL
* `ntag` set initial nTag value
* `clientid` set the client-id default header

### Passing path parameters

```js
import { get } from 'nordnet-next-api';

get('https://api.test.nordnet.se/next/2/accounts/{accno}', { accno: 123456789 })
  .then(({ status, data }) => console.log(status, data));
```

### Passing query parameters

```js
import { get } from 'nordnet-next-api';

get('https://api.test.nordnet.se/next/2/news?days={days}', { days: 0 })
  .then(({ status, data }) => console.log(status, data));
```

### Passing POST parameters

```js
import { post } from 'nordnet-next-api';

post('https://api.test.nordnet.se/next/2/user/{key}', { key: 'foo', value: { bar: 'bar' }})
  .then(({ status, data }) => console.log(status, data));
```

### Passing additional headers

```js
import { get } from 'nordnet-next-api';

get('https://api.test.nordnet.se/next/2/markets/{market_id}', { market_id: 11 }, { 'Accept-Language': 'sv' })
  .then(({ status, data }) => console.log(status, data));
```

See tests under `src/__tests__` for more examples.

### Passing custom HTTP agent
```js
import http from 'http';
import { get } from 'nordnet-next-api';

const agent = new http.Agent();

get('https://api.test.nordnet.se/next/2/markets/{market_id}', { market_id: 11 }, { 'Accept-Language': 'sv' }, { agent })
  .then(({ status, data }) => console.log(status, data));
```

### Setting up default HTTP agent
```js
import http from 'http';
import api from 'nordnet-next-api';

const agent = new http.Agent();

api.setConfig({ agent });

get('https://api.test.nordnet.se/next/2/markets/{market_id}', { market_id: 11 }, { 'Accept-Language': 'sv' })
  .then(({ status, data }) => console.log(status, data));
```

## Example projects

nordnet-next-api is distributed with a two simple example projects.

Before proceeding, install dependencies and build the project:

```
npm install
npm run build
```

Run the client side example:

```
cd examples/client
npm install
npm start
```

Run the server side example:

```
cd examples/server
npm install
npm start
```

## License

This Open Source project released by Nordnet is licensed under the MIT license.


[api]: https://api.test.nordnet.se/
[api-docs]: https://api.test.nordnet.se/api-docs/index.html

[response]: https://developer.mozilla.org/en-US/docs/Web/API/Response
[fetch-api]: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

[npm-url]: https://npmjs.org/package/nordnet-next-api
[npm-image]: https://img.shields.io/npm/v/nordnet-next-api.svg

[travis-url]: https://travis-ci.org/nordnet/nordnet-next-api
[travis-image]: https://travis-ci.org/nordnet/nordnet-next-api.svg?branch=master

[depstat-url]: https://david-dm.org/nordnet/nordnet-next-api
[depstat-image]: https://david-dm.org/nordnet/nordnet-next-api.svg


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 5.1.2  (2018-01-30)

* Allow minor releases of isomorphic-fetch (^2.1.1)
* Add yarn support

## 5.1.1 (2017-09-18)

* Support `0` as a valid path parameter value. Example:

```
url: '/api-url/{score}',
params: {
  score,
},
```

## 5.1.0

* Support PDF
* Supports lowercase ntag and clientid as options. Still backward compatible.

## 5.0.0

* BREAKING: remove Promise polyfill

# Previous releases

TODO: write about past releases, if needed


src/index.js
import 'isomorphic-fetch';

const HTTP_NO_CONTENT = 204;
const HTTP_BAD_REQUEST = 400;
const regUrlParam = /{([\s\S]+?)}/g;

const defaultHeaders = {
  accept: 'application/json',
};

function postDefaultHeaders() {
  return Object.assign(
    { 'content-type': 'application/x-www-form-urlencoded' },
    defaultHeaders
  );
}

const state = {
  ntag: 'NO_NTAG_RECEIVED_YET',
};

const credentials = 'include';

const config = {};
const configKeys = ['root', 'agent'];

export function setConfig(options = {}) {
  // HTTP headers should be lowercase, but we support camelCase for backwards compatibility
  const lowerCaseOptions = sanitizeHeaders(options);
  configKeys.forEach(key => config[key] = options[key]);

  if (lowerCaseOptions.ntag) {
    state.ntag = lowerCaseOptions.ntag;
  }
  if (lowerCaseOptions.clientid) {
    defaultHeaders['client-id'] = lowerCaseOptions.clientid;
  }
}

export function get(url, params = {}, headers = {}, options = {}) {
  const fetchOptions = Object.assign({
    url,
    params,
    headers,
    method: 'get',
  }, options);

  return httpFetch(fetchOptions);
}

export function post(url, params = {}, headers = {}, options = {}) {
  const fetchOptions = Object.assign({
    url,
    params,
    headers,
    method: 'post',
  }, options);

  return httpFetch(fetchOptions);
}

export function postJson(url, params = {}, headers = {}, options = {}) {
  return post(url, params, Object.assign(headers, { 'Content-type': 'application/json' }, options));
}

export function put(url, params = {}, headers = {}, options = {}) {
  const fetchOptions = Object.assign({
    url,
    params,
    headers,
    method: 'put',
  }, options);

  return httpFetch(fetchOptions);
}

export function putJson(url, params = {}, headers = {}, options = {}) {
  return put(url, params, Object.assign(headers, { 'Content-type': 'application/json' }, options));
}

export function del(url, params = {}, headers = {}, options = {}) {
  const fetchOptions = Object.assign({
    url,
    params,
    headers,
    method: 'delete',
  }, options);

  return httpFetch(fetchOptions);
}

export default {
  get,
  post,
  postJson,
  put,
  putJson,
  del,
  setConfig,
};

function httpFetch(options) {
  if (!options.url) { // @TODO should check type
    return Promise.reject(new Error(`Invalid url, got \`${options.url}\``));
  }

  if (isNotValidPath(options.url, options.params)) { // @TODO should check types
    return Promise.reject(new Error(`Params object doesn't have all required keys for url.
      Got url \`${options.url}\` and params \`${JSON.stringify(options.params)}\``));
  }

  const path = buildPath(options.url, options.params);
  const params = omit(options.params, getPathParams(options.url));

  const query = hasQuery(options.method) ? buildParams(params) : undefined;
  const headers = buildHeaders(options.method, options.headers);
  const body = buildBody(options.method, params, headers);

  const fetchUrl = buildUrl(path, query);

  const fetchParams = {
    headers,
    credentials,
    body,
    method: options.method,
  };

  if (options.agent) {
    Object.assign(fetchParams, { agent: options.agent });
  } else if (config.agent) {
    Object.assign(fetchParams, { agent: config.agent });
  }

  return fetch(fetchUrl, fetchParams)
    .then(validateStatus)
    .then(saveNTag)
    .then(processResponse);
}

function validateStatus(response) {
  if (response.status < HTTP_BAD_REQUEST) {
    return response;
  }

  return toErrorResponse(response);
}

function toErrorResponse(response) {
  return parseContent(response).then(res => Promise.reject(res));
}

function saveNTag(response) {
  state.ntag = response.headers.get('ntag') || state.ntag;
  return response;
}

function processResponse(response) {
  if (response.status === HTTP_NO_CONTENT) {
    return { response, status: response.status };
  }

  return parseContent(response);
}

function parseContent(response) {
  const contentType = response.headers.get('Content-type');
  let method = isJSON(contentType) ? 'json' : 'text';

  if (isBlob(contentType)) {
    method = 'blob';
  }

  return response[method]()
    .then(data => ({ response, data, status: response.status }))
    .catch(error => Promise.reject({
      response,
      status: response.status,
      error: new Error(`fetch unable to parse input, most likely API responded with empty reponse.
        Original Error: ${error}`),
    }));
}

function isJSON(contentType) {
  return contains('application/json')(contentType);
}

function isBlob(contentType) {
  return contains('application/pdf')(contentType);
}

function getPathParams(url) {
  const keys = url.match(regUrlParam) || [];
  return keys.map(key => key.replace(/({|})/g, ''));
}

function buildUrl(path, query = '') {
  const queryParams = query.length ? query.join('&') : '';
  const pathContainsQuery = path.indexOf('?') !== -1;
  const pathContainsProtocol = !!(path.match(/^http(s)?:\/\//));

  let root = '';
  if (!pathContainsProtocol && typeof config.root !== 'undefined') {
    root = config.root;
  }

  let delimiter = '';
  if (pathContainsQuery && queryParams) {
    delimiter = '&';
  } else if (queryParams) {
    delimiter = '?';
  }

  return `${root}${path}${delimiter}${queryParams}`;
}

function isNotValidPath(url, params = {}) {
  return !!(url.match(regUrlParam) || [])
    .map(key => key.replace(/({|})/g, ''))
    .find(key => !params[key] && params[key] !== 0);
}

function buildPath(url, params) {
  return url.replace(regUrlParam, matchParams(params));
}

function matchParams(params) {
  return function matchParamKeyValue(match, key) {
    return uriEncode(params[key]);
  };
}

function buildParams(params = {}) {
  return Object.keys(params).map(key => encodeURIComponent(key) + '=' + uriEncode(params[key]));
}

function buildHeaders(method, headers) {
  return Object.assign({ ntag: state.ntag }, getDefaultMethodHeaders(method), sanitizeHeaders(headers));
}

function getDefaultMethodHeaders(method) {
  return method === 'post' || method === 'put' ? postDefaultHeaders() : defaultHeaders;
}

function sanitizeHeaders(obj) {
  return Object.keys(obj).reduce(keyToLowerCase(obj), {});
}

function keyToLowerCase(obj) {
  return (accumulator, key) => {
    accumulator[key.toLowerCase()] = obj[key];
    return accumulator;
  };
}

function buildBody(method, params, headers) {
  if (!hasBody(method)) {
    return undefined;
  }

  return isJsonContentType(headers) ? JSON.stringify(params) : buildParams(params).join('&');
}

function isJsonContentType(headers) {
  const contentType = Object.keys(headers).find(contains('content-type'));
  return isJSON(headers[contentType]);
}

function uriEncode(value) {
  let encoded;
  if (Array.isArray(value)) {
    encoded = value.join(',');
  } else if (isPlainObject(value)) {
    encoded = JSON.stringify(value);
  } else {
    encoded = value;
  }

  return encodeURIComponent(encoded);
}

function hasQuery(method) {
  return method === 'get' || method === 'delete';
}

function hasBody(method) {
  return method === 'post' || method === 'put';
}

function contains(string) {
  return function (value) {
    return !!value && value.toLowerCase().indexOf(string) !== -1;
  };
}

function omit(source, props) {
  return Object.keys(source).reduce(omitKey(source, props), {});
}

function omitKey(source, props) {
  return (accumulator, key) => {
    if (props.indexOf(key) === -1) {
      accumulator[key] = source[key];
    }

    return accumulator;
  };
}

function isPlainObject(obj) {
  return obj !== null && typeof obj === 'object';
}
